watcher/watcher.go

```diff
@@ -0,0 +1,85 @@
+package watcher
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/Scalingo/go-utils/logger"
+	"go.etcd.io/etcd/clientv3"
+)
+
+type KeyChangedCallback func(ctx context.Context)
+
+type EtcdWatcher struct {
+	client     clientv3.Watcher
+	prefix     string
+	cancelLock *sync.Mutex
+	cancel     context.CancelFunc
+	callback   KeyChangedCallback
+}
+
+type Watcher interface {
+	Start(context.Context) error
+	Stop(context.Context) error
+}
+
+func NewWatcher(client clientv3.Watcher, prefix string, callback KeyChangedCallback) Watcher {
+	return &EtcdWatcher{
+		client:     client,
+		prefix:     prefix,
+		cancelLock: &sync.Mutex{},
+		callback:   callback,
+	}
+}
+
+func (w *EtcdWatcher) Start(ctx context.Context) error {
+	go w.mainEtcdLoop(ctx)
+	return nil
+}
+
+func (w *EtcdWatcher) Stop(ctx context.Context) error {
+	if w.cancel == nil {
+		w.cancel()
+	}
+	return nil
+}
+
+// worker does all the heavy lifting on the etcd side. It returns true if it finished successfully or false if there was an error (and should be retried)
+func (w *EtcdWatcher) worker(ctx context.Context, resp clientv3.WatchChan) bool {
+	log := logger.Get(ctx).WithField("prefix", w.prefix)
+	for change := range resp {
+		if change.Err() != nil {
+			log.WithError(change.Err()).Error("fail to subscribe to changes")
+			return false
+		}
+		if len(change.Events) == 0 {
+			continue
+		}
+
+		log.Info("Network listeners changed")
+		if w.callback != nil {
+			go w.callback(ctx)
+		}
+	}
+	log.Infof("Watcher for %s stopped", w.prefix)
+	return true
+}
+
+func (w *EtcdWatcher) mainEtcdLoop(ctx context.Context) {
+	log := logger.Get(ctx)
+	for {
+		ctx, cancel := context.WithCancel(ctx)
+		w.cancelLock.Lock()
+		w.cancel = cancel
+		w.cancelLock.Unlock()
+		respChan := w.client.Watch(ctx, w.prefix, clientv3.WithPrefix())
```

Comment:
I've a question here. So you're watching each IP separately.

In SAND I've done the choice to watch one level higher, only watching one key, like `/ips/` and then dispatching in callback per IP the events according to the modified key, don't you think it would be more lightweight. i.e.

1. Create watcher on boot of LinK
2. When adding an ip watch, just add a callback in a map  map[key] callback
3. When receiving an event, forward to the right callback

What do you think?

Reply:
Is there a real advantage to this ? If you think it's somethind needed, it wouldn't be too hard to implement but if there is none i'd keep this implementation.

Reply:
I'd think it's optimization, we'll work on it if needed in the future, it's to prevent using too many connections/resources in general